django/core/cache/backends/locmem.py

```diff
@@ -13,117 +13,97 @@
-        with self._lock.writer():
+        with self._lock:
             if self._has_expired(key):
                 self._set(key, pickled, timeout)
                 return True
             return False
 
-    def get(self, key, default=None, version=None, acquire_lock=True):
+    def get(self, key, default=None, version=None):
         key = self.make_key(key, version=version)
         self.validate_key(key)
-        pickled = None
-        with (self._lock.reader() if acquire_lock else dummy()):
+        with self._lock:
             if not self._has_expired(key):
                 pickled = self._cache[key]
-        if pickled is not None:
-            try:
-                return pickle.loads(pickled)
-            except pickle.PickleError:
+                self._cache.move_to_end(key, last=False)
+            else:
+                self._discard(key)
                 return default
+        return pickle.loads(pickled)
 
-        with (self._lock.writer() if acquire_lock else dummy()):
-            try:
-                del self._cache[key]
-                del self._expire_info[key]
-            except KeyError:
-                pass
-            return default
+    def set(self, key, value, timeout=DEFAULT_TIMEOUT, version=None):
+        key = self.make_key(key, version=version)
+        self.validate_key(key)
+        pickled = pickle.dumps(value, pickle.HIGHEST_PROTOCOL)
+        with self._lock:
+            self._set(key, pickled, timeout)
 
     def _set(self, key, value, timeout=DEFAULT_TIMEOUT):
         if len(self._cache) >= self._max_entries:
             self._cull()
         self._cache[key] = value
+        self._cache.move_to_end(key, last=False)
         self._expire_info[key] = self.get_backend_timeout(timeout)
 
-    def set(self, key, value, timeout=DEFAULT_TIMEOUT, version=None):
-        key = self.make_key(key, version=version)
-        self.validate_key(key)
-        pickled = pickle.dumps(value, pickle.HIGHEST_PROTOCOL)
-        with self._lock.writer():
-            self._set(key, pickled, timeout)
+    def _cull(self):
+        if self._cull_frequency == 0:
+            self.clear()
+        else:
+            count = len(self._cache) // self._cull_frequency
+            for i in range(count):
+                key, _ = self._cache.popitem()
+                del self._expire_info[key]
 
     def incr(self, key, delta=1, version=None):
-        with self._lock.writer():
-            value = self.get(key, version=version, acquire_lock=False)
+        with self._lock:
+            value = self.get(key, version=version)
             if value is None:
                 raise ValueError("Key '%s' not found" % key)
             new_value = value + delta
             key = self.make_key(key, version=version)
             pickled = pickle.dumps(new_value, pickle.HIGHEST_PROTOCOL)
             self._cache[key] = pickled
+            self._cache.move_to_end(key, last=False)
```

Comment:
Commenting this line has no effect on the tests. (The call to `self.get()` above already applies the move.)

Reply:
I updated the code to remove the dependence on get() from incr(). Now the call is required.